### ISSUE TITLE
Multiple loss.backward() for each minibatch during training

### ISSUE THREAD

**Issue by itzsid** [filipradenovic/cnnimageretrieval-pytorch#20]
Hi @filipradenovic,

Sorry for a trivial question but I'm not aware how PyTorch stores gradient for multiple calls of ```loss.backward()``` during before calling ```optimizer.step()```. 
1. Does it sum all the gradients or store an average? 
2. Is there a reason you did call loss.backward() multiple times and not summing up all the losses and calling loss.backward() once?  
3. Also, is there a reason for computing loss for each tuple separately in minibatch for training and computing loss for the whole minibatch together during validation?

Thanks!

**Comment by filipradenovic** [comment]
1. As far as I know it sums all the gradients, take a look at [this discussion](https://discuss.pytorch.org/t/pytorch-gradients/884) for more insights about PyTorch and gradients. A really important thing is to do `optimizer.zero_grad()` before the loop in which you do `loss.backward()`, because the gradients need to be cleared before accumulating them for a new batch.
2. First of all, both implementations would yield the same result. But yes, there is a reason why we do it the way we do. It is a trade-off between performance and memory usage. As it is implemented right now, it is the most memory efficient (so that you can do training with larger images as well, if necessary). Doing only one `loss.backward()` per `optimizer.step()` would be the most memory hungry, but also the fastest. Code can be simply adjusted if you want to do that. You can also implement it to split each batch into smaller sub-batches, so that they nearly fill up the memory.
3. During validation there is no backward pass or gradient computation, it is a simple evaluation of the network so it does not require a lot of memory. Because of that, we do not need to split the loss computations, we can simply compute it for the whole batch. 

Hope this helps!

**Comment by itzsid** [comment]
Thanks for the quick replies @filipradenovic. This is helpful.
